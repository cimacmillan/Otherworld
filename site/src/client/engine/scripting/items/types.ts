import { GameItem } from "../../../resources/manifests/Items";

export interface Item {
    id: GameItem;
    spriteIcon: string;
    stackable: boolean;
    name: string;
    value?: number;
    behaviours: ItemComponent[];
    category: ItemCategory;
    description: string;
}

export enum ItemCategory {
    CONSUMABLE = "Consumable",
    CRAFTING = "Crafting",
    PRECIOUS = "Precious",
    KEY = "Key",
}

export interface ItemMetadata {
    item: Item;
    count: number;
}

export interface Inventory {
    items: ItemMetadata[];
}

export const getEmptyInventory = (): Inventory => ({
    items: [],
});

export type ItemComponent = ItemHealsPlayer | MakesNoiseWhenConsumed;

export interface ItemHealsPlayer {
    type: ItemComponentType.HEALS_PLAYER;
    amount: number;
}

export interface MakesNoiseWhenConsumed {
    type: ItemComponentType.MAKES_NOISE_WHEN_CONSUMED;
}

export enum ItemComponentType {
    HEALS_PLAYER = "HEALS_PLAYER",
    MAKES_NOISE_WHEN_CONSUMED = "MAKES_NOISE_WHEN_CONSUMED",
}
