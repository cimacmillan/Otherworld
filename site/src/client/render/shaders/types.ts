
interface UniformPositions {
    [key: string]: WebGLUniformLocation;
}

interface AttributePositions {
    [key: string]: number;
}

interface CompiledShader {
    shaderId: WebGLProgram;
    uniform: UniformPositions;
    attribute: AttributePositions;
}
