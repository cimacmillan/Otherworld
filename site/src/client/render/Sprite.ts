import { ScreenBuffer, DepthBuffer } from ".";
import { Camera, Sprite } from "../types";
import { vec_sub, vec_add, vec_rotate, vec_distance, clipToRange, swapSort } from "../util/math";
import { basicShader } from "./Shader";

function drawSprite(screen: ScreenBuffer, depth_buffer: DepthBuffer, camera: Camera, sprite: Sprite) {
    const projectPosition = sprite.projectPosition!;

    const diff = vec_sub(sprite.position, {x: 5, y: 5});
    sprite.position = vec_add(vec_rotate(diff, 0.04 / Math.pow(vec_distance(diff), 2) ), {x: 5, y: 5});

    // Angle of 0 is looking up? 
    const distance = -projectPosition.y;

    if (distance < camera.clip_depth) {
        return;
    }

    const projectionMultiplier = (camera.focal_length / distance) * camera.space_to_pixel;
    
    let width = sprite.size.x * projectionMultiplier; 
    let height = sprite.size.y * projectionMultiplier;
    
    let x = (projectPosition.x * projectionMultiplier) + (0.5 * screen.width);
    let y = (-sprite.height + camera.height) * projectionMultiplier + (0.5 * screen.height);

    let x1 = clipToRange(Math.floor(x - width/2), 0, screen.width-1);
    let x2 = clipToRange(Math.floor(x + width/2), 0, screen.width-1);
    let y1 = clipToRange(Math.floor(y - height/2), 0, screen.height-1);
    let y2 = clipToRange(Math.floor(y + height/2), 0, screen.height-1);
    
    for (let xPixel = x1; xPixel < x2; xPixel++) {
        for (let yPixel = y1; yPixel < y2; yPixel++) {
            screen.putPixelColour(xPixel, yPixel, basicShader(sprite.colour));
            depth_buffer.setDistance(xPixel, yPixel, distance);
        }
    }

}

export function drawSprites(screen: ScreenBuffer, depth_buffer: DepthBuffer, camera: Camera, sprites: Sprite[]) {
    sprites.forEach((sprite) => {
        sprite.projectPosition = vec_rotate(vec_sub(sprite.position, camera.position), -camera.angle)
    })
    
    swapSort(sprites, (spriteA, spriteB) => (spriteA.projectPosition!.y < spriteB.projectPosition!.y));

    sprites.forEach((sprite) => {
        drawSprite(screen, depth_buffer, camera, sprite);
    });
}