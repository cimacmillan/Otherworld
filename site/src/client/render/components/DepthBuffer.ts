import { fill } from "../../util/math/Array";

export class DepthBuffer {

    data: number[];
    width: number;
    height: number;

    constructor(width: number, height: number) {
        this.data = Array(width * height).fill(0);
        this.width = width;
        this.height = height;
    }

    isCloser(x: number, y: number, distance: number) {
        let index = (y * this.width) + x;
        return 1.0 / distance > this.data[index];
    }

    setDistance(x: number, y: number, distance: number) {
        let index = (y * this.width) + x;
        this.data[index] = 1.0 / distance;
    }

    reset() {
        fill(this.data, 0);
    }
}
