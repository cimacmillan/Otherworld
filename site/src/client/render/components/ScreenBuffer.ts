import { Colour } from "../../types";
import { fillPattern } from "../../util/math/SyncedArray";

export class ScreenBuffer {
  // public image_data: ImageData;
  private gl: WebGLRenderingContext;
  public width: number;
  public height: number;

  constructor(gl: WebGLRenderingContext, width: number, height: number) {
    this.gl = gl;
    this.width = width;
    this.height = height;
  }

  public getOpenGL() {
    return this.gl;
  }

  // public putPixel(x: number, y: number, red: number, green: number, blue: number, alpha: number) {
  //     const pixelindex = (y * this.width + x) * 4;
  //     this.image_data.data[pixelindex] = red;
  //     this.image_data.data[pixelindex + 1] = green;
  //     this.image_data.data[pixelindex + 2] = blue;
  //     this.image_data.data[pixelindex + 3] = alpha;
  // }

  // public putPixelColour(x: number, y: number, colour: Colour, depth: number, farClip: number, backgroundColour: Colour) {

  //     let colourMult = (depth / farClip);
  //     colourMult *= colourMult;

  //     // Breaking the rules but it's faster
  //     this.putPixel(x, y,
  //         ~~(colour.r + colourMult * (backgroundColour.r - colour.r)),
  //         ~~(colour.g + colourMult * (backgroundColour.g - colour.g)),
  //         ~~(colour.b + colourMult * (backgroundColour.b - colour.b)),
  //         ~~(colour.a + colourMult * (backgroundColour.a - colour.a)),
  //         );
  // }

  // public fillBackground(red: number, green: number, blue: number, alpha: number) {
  //     fillPattern(this.image_data.data, [red, green, blue, alpha]);
  // }

  // public reset(colour: Colour) {
  //     this.fillBackground(colour.r, colour.g, colour.b, 255);
  // }
}
