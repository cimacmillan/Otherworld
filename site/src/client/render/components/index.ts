export * from "./CanvasComponent"
export * from "./DepthBuffer"
export * from "./ScreenBuffer"