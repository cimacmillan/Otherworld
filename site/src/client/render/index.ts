export * from "./components/CanvasComponent";
export * from "./components/DepthBuffer";
export * from "./components/ScreenBuffer";
export * from "./Render";
export * from "./Planes";
