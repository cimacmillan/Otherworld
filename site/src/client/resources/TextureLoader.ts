import { SpriteSheet } from "./SpriteSheet";
import { SpriteSheetManifestJson } from "./Types";

export function loadImage(url: string): Promise<HTMLImageElement> {
    return new Promise<HTMLImageElement>((resolve) => {
        const image = new Image();
        image.onload = function () {
            resolve(image);
        };
        image.onerror = (error) => {
            console.log(`Error loading image ${url}`, error);
        };
        image.src = url;
    });
}

export function loadJson<T>(url: string): Promise<T> {
    return fetch(url).then(response => response.json())
}

export function loadTextureFromImage(
    gl: WebGLRenderingContext,
    image: HTMLImageElement
): WebGLTexture {
    const texture = gl.createTexture();
    gl.bindTexture(gl.TEXTURE_2D, texture);
    gl.bindTexture(gl.TEXTURE_2D, texture);
    gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, gl.RGBA, gl.UNSIGNED_BYTE, image);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.NEAREST);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.NEAREST);
    return texture;
}

export function loadTextureFromURL(
    gl: WebGLRenderingContext,
    url: string
): Promise<WebGLTexture> {
    return loadImage(url).then((image: HTMLImageElement) => {
        return loadTextureFromImage(gl, image);
    });
}

export async function loadSpriteSheet(
    gl: WebGLRenderingContext,
    url: string
): Promise<[SpriteSheet, SpriteSheetManifestJson]> {
    const imageUrl = url + ".png";
    const manifestUrl = url + ".json";

    const image = await loadImage(imageUrl);
    const json = await loadJson<SpriteSheetManifestJson>(manifestUrl);

    const texture = loadTextureFromImage(gl, image);
    return [new SpriteSheet(image.width, image.height, texture, image), json];
}

export function getImageData(img: HTMLImageElement) {
    const canvas = document.createElement("canvas");
    canvas.width = img.width;
    canvas.height = img.height;
    canvas.getContext("2d").drawImage(img, 0, 0, img.width, img.height);
    const data = canvas
        .getContext("2d")
        .getImageData(0, 0, img.width, img.height);
    return data;
}

export async function loadImageData(url: string) {
    const img = await loadImage(url);
    const canvas = document.createElement("canvas");
    canvas.width = img.width;
    canvas.height = img.height;
    canvas.getContext("2d").drawImage(img, 0, 0, img.width, img.height);
    const data = canvas
        .getContext("2d")
        .getImageData(0, 0, img.width, img.height);
    return data;
}
