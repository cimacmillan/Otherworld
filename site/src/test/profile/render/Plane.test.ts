import { drawPlanes } from "../../../client/render";
import { DepthBuffer } from "../../../client/render/components/DepthBuffer";
import { ScreenBuffer } from "../../../client/render/components/ScreenBuffer";
import { drawSprite, drawSprites} from "../../../client/render/Sprite";
import { Camera, Plane, SpriteSheet, Texture } from "../../../client/types";
import { convertToFastTexture } from "../../../client/util/loader/TextureLoader";
import { profile, SAMPLE_LIL, SAMPLE_MEDIUM_RARE, SAMPLE_SMA } from "../ProfileUnit";
import { mockDepthBuffer, mockScreenBuffer } from "../util";

const randomTexture = (width: number, height: number): Texture => {
    const data = new Uint8ClampedArray(width * height * 4);
    return {data: {data, width, height}, width, height};
};

const MANY_PLANES = 5000;

describe("Plane Profile", () => {
    let camera: Camera;

    beforeEach(() => {
        camera = {
            position: { x: 5.0, y: 5.0 },
            angle: -Math.PI/2 + 0.1,
            focal_length: 1,
            height: 1,
            x_view_window: 1,
            y_view_window: 1,
            clip_depth: 0.1,
            far_clip_depth: 20
        };
    });

    describe("Many planes", () => {

        let floor: Plane;
        let ceiling: Plane;
        const colour = {r: 255, g: 255, b: 255, a: 255};

        beforeEach(() => {
            const texture = convertToFastTexture(randomTexture(64, 64));
            const spritesheet: SpriteSheet = {
                data: [[texture]],
                width: 1,
                height: 1,
            };

            floor = {
                height: 0,
                start: {
                    x: 0,
                    y: 0,
                },
                end: {
                    x: 10,
                    y: 10,
                },
                spritesheet,
            };

            ceiling = {
                height: 2,
                start: {
                    x: 0,
                    y: 0,
                },
                end: {
                    x: 10,
                    y: 10,
                },
                spritesheet,
            };

        });

        test("rendering planes", () => {
            profile(MANY_PLANES, () => drawPlanes(mockScreenBuffer, mockDepthBuffer, camera, colour, floor, ceiling));
        });

    });

});
