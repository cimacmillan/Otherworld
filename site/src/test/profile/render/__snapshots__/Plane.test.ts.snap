// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plane Profile Many planes rendering planes 1`] = `5438`;
